File: server/utils/usage_limits.py
"""
Usage limits enforcement for Briefly Cloud
Handles tier-based limits and usage tracking
"""

from typing import Dict, Optional, Tuple
from fastapi import HTTPException
import logging
from datetime import datetime
from supabase import Client

logger = logging.getLogger(__name__)

# Updated tier limits - cost-conscious values
TIER_LIMITS = {
    'free': {
        'documents': 10,
        'chat_messages': 100,
        'api_calls': 1000,
        'storage_bytes': 104857600,  # 100MB
    },
    'pro': {
        'documents': 1000,
        'chat_messages': 1000,
        'api_calls': 10000,
        'storage_bytes': 10737418240,  # 10GB
    },
    'pro_byok': {
        'documents': 10000,
        'chat_messages': 5000,
        'api_calls': 50000,
        'storage_bytes': 107374182400,  # 100GB
    }
}

class UsageLimitError(HTTPException):
    """Custom exception for usage limit violations"""
    def __init__(self, limit_type: str, current: int, limit: int, tier: str):
        self.limit_type = limit_type
        self.current = current
        self.limit = limit
        self.tier = tier
        
        detail = {
            "error": "usage_limit_exceeded",
            "message": f"You have exceeded your {limit_type} limit for the {tier} tier",
            "current_usage": current,
            "limit": limit,
            "tier": tier,
            "upgrade_required": tier == 'free'
        }
        
        super().__init__(status_code=429, detail=detail)

async def get_user_usage(supabase: Client, user_id: str) -> Optional[Dict]:
    """Get current user usage statistics"""
    try:
        result = supabase.table('users').select(
            'subscription_tier, documents_uploaded, documents_limit, '
            'chat_messages_count, chat_messages_limit, '
            'api_calls_count, api_calls_limit, '
            'storage_used_bytes, storage_limit_bytes, '
            'usage_reset_date'
        ).eq('id', user_id).single().execute()
        
        return result.data if result.data else None
    except Exception as e:
        logger.error(f"Error fetching user usage for {user_id}: {e}")
        return None

async def check_usage_limit(
    supabase: Client, 
    user_id: str, 
    limit_type: str, 
    increment: int = 1
) -> Tuple[bool, Dict]:
    """
    Check if user can perform an action without exceeding limits
    
    Args:
        supabase: Supabase client
        user_id: User UUID
        limit_type: 'documents', 'chat_messages', 'api_calls', or 'storage'
        increment: Amount to increment (default 1)
    
    Returns:
        Tuple of (within_limits, usage_data)
    """
    usage = await get_user_usage(supabase, user_id)
    if not usage:
        return False, {}
    
    tier = usage['subscription_tier']
    
    # Map limit types to database columns
    limit_mapping = {
        'documents': ('documents_uploaded', 'documents_limit'),
        'chat_messages': ('chat_messages_count', 'chat_messages_limit'),
        'api_calls': ('api_calls_count', 'api_calls_limit'),
        'storage': ('storage_used_bytes', 'storage_limit_bytes')
    }
    
    if limit_type not in limit_mapping:
        logger.error(f"Invalid limit type: {limit_type}")
        return False, usage
    
    current_col, limit_col = limit_mapping[limit_type]
    current_usage = usage[current_col] or 0
    limit = usage[limit_col] or 0
    
    # Check if adding increment would exceed limit
    would_exceed = (current_usage + increment) > limit
    
    return not would_exceed, {
        'tier': tier,
        'current': current_usage,
        'limit': limit,
        'would_exceed': would_exceed,
        'remaining': max(0, limit - current_usage)
    }

async def enforce_usage_limit(
    supabase: Client, 
    user_id: str, 
    limit_type: str, 
    increment: int = 1
) -> Dict:
    """
    Enforce usage limits - raise exception if exceeded
    
    Args:
        supabase: Supabase client
        user_id: User UUID
        limit_type: 'documents', 'chat_messages', 'api_calls', or 'storage'
        increment: Amount to increment (default 1)
    
    Returns:
        Usage data dictionary
        
    Raises:
        UsageLimitError: If limit would be exceeded
    """
    within_limits, usage_data = await check_usage_limit(
        supabase, user_id, limit_type, increment
    )
    
    if not within_limits:
        raise UsageLimitError(
            limit_type=limit_type,
            current=usage_data['current'],
            limit=usage_data['limit'],
            tier=usage_data['tier']
        )
    
    return usage_data

async def increment_usage_counter(
    supabase: Client,
    user_id: str,
    event_type: str,
    resource_count: int = 1,
    event_data: Dict = None
) -> bool:
    """
    Increment usage counter using database function
    
    Args:
        supabase: Supabase client
        user_id: User UUID
        event_type: 'document_upload', 'chat_message', 'api_call'
        resource_count: Amount to increment
        event_data: Additional event metadata
    
    Returns:
        Success boolean
    """
    try:
        result = supabase.rpc('increment_usage', {
            'p_user_id': user_id,
            'p_event_type': event_type,
            'p_resource_count': resource_count,
            'p_event_data': event_data or {}
        }).execute()
        
        return result.data if result.data else False
    except Exception as e:
        logger.error(f"Error incrementing usage for {user_id}: {e}")
        return False

async def check_and_increment_usage(
    supabase: Client,
    user_id: str,
    limit_type: str,
    event_type: str,
    increment: int = 1,
    event_data: Dict = None
) -> Dict:
    """
    Check limits and increment usage in one operation
    
    Args:
        supabase: Supabase client
        user_id: User UUID
        limit_type: 'documents', 'chat_messages', 'api_calls'
        event_type: 'document_upload', 'chat_message', 'api_call'
        increment: Amount to increment
        event_data: Additional event metadata
    
    Returns:
        Usage data dictionary
        
    Raises:
        UsageLimitError: If limit would be exceeded
    """
    # First check if within limits
    usage_data = await enforce_usage_limit(supabase, user_id, limit_type, increment)
    
    # If within limits, increment the counter
    success = await increment_usage_counter(
        supabase, user_id, event_type, increment, event_data
    )
    
    if not success:
        logger.error(f"Failed to increment usage counter for {user_id}")
    
    return usage_data

def get_usage_warning_threshold(tier: str, limit_type: str) -> int:
    """Get the threshold at which to warn users about approaching limits"""
    if tier not in TIER_LIMITS or limit_type not in TIER_LIMITS[tier]:
        return 0
    
    limit = TIER_LIMITS[tier][limit_type]
    
    # Warn at 80% of limit
    return int(limit * 0.8)

def format_storage_size(bytes_size: int) -> str:
    """Format storage size in human-readable format"""
    for unit in ['B', 'KB', 'MB', 'GB', 'TB']:
        if bytes_size < 1024.0:
            return f"{bytes_size:.1f} {unit}"
        bytes_size /= 1024.0
    return f"{bytes_size:.1f} PB"

def get_upgrade_message(tier: str, limit_type: str) -> str:
    """Get appropriate upgrade message based on current tier and limit type"""
    if tier == 'free':
        return "Upgrade to Pro for higher limits and better performance."
    elif tier == 'pro':
        return "Consider Pro BYOK for even higher limits with your own API keys."
    else:
        return "You're on our highest tier. Contact support for enterprise options."

async def reset_monthly_usage(supabase: Client) -> bool:
    """Reset monthly usage counters for all users"""
    try:
        result = supabase.rpc('reset_monthly_usage').execute()
        logger.info("Monthly usage reset completed")
        return True
    except Exception as e:
        logger.error(f"Error resetting monthly usage: {e}")
        return False

async def get_usage_analytics(supabase: Client, days: int = 30) -> Dict:
    """Get usage analytics for the specified number of days"""
    try:
        result = supabase.table('usage_analytics').select('*').limit(days * 10).execute()
        return result.data if result.data else []
    except Exception as e:
        logger.error(f"Error fetching usage analytics: {e}")
        return []

File: briefly-cloud-nextjs/src/app/lib/usage-limits.ts
import { supabaseApp } from './supabase-clients'
import type { User } from './supabase'
import { getUserById, updateUser } from './supabase'

// Updated tier limits (migrated from Python)
export const TIER_LIMITS = {
  free: {
    documents: 25,
    chat_messages: 100,
    api_calls: 1000,
    storage_bytes: 104857600, // 100MB
  },
  pro: {
    documents: 500,
    chat_messages: 400,
    api_calls: 10000,
    storage_bytes: 1073741824, // 1GB
  },
  pro_byok: {
    documents: 5000,
    chat_messages: 2000,
    api_calls: 50000,
    storage_bytes: 10737418240, // 10GB
  }
} as const

export type SubscriptionTier = keyof typeof TIER_LIMITS
export type LimitType = keyof typeof TIER_LIMITS[SubscriptionTier]

export class UsageLimitError extends Error {
  constructor(
    public limitType: string,
    public current: number,
    public limit: number,
    public tier: string,
    public upgradeRequired: boolean = false
  ) {
    super(`Usage limit exceeded for ${limitType}: ${current}/${limit} (${tier} tier)`)
    this.name = 'UsageLimitError'
  }

  toJSON() {
    return {
      error: 'usage_limit_exceeded',
      message: `You have exceeded your ${this.limitType} limit for the ${this.tier} tier`,
      current_usage: this.current,
      limit: this.limit,
      tier: this.tier,
      upgrade_required: this.upgradeRequired
    }
  }
}

export interface UsageData {
  tier: SubscriptionTier
  current: number
  limit: number
  would_exceed: boolean
  remaining: number
}

export async function getUserUsage(userId: string): Promise<User | null> {
  try {
    return await getUserById(userId)
  } catch (error) {
    console.error(`Error fetching user usage for ${userId}:`, error)
    return null
  }
}

export async function checkUsageLimit(
  userId: string,
  limitType: LimitType,
  increment: number = 1
): Promise<{ withinLimits: boolean; usageData: UsageData }> {
  const user = await getUserUsage(userId)
  if (!user) {
    return {
      withinLimits: false,
      usageData: {
        tier: 'free',
        current: 0,
        limit: 0,
        would_exceed: true,
        remaining: 0
      }
    }
  }

  const tier = user.subscription_tier as SubscriptionTier
  const tierLimits = TIER_LIMITS[tier]

  // Map limit types to database columns
  const limitMapping = {
    documents: {
      current: user.documents_uploaded || 0,
      limit: user.documents_limit || tierLimits.documents
    },
    chat_messages: {
      current: user.chat_messages_count || 0,
      limit: user.chat_messages_limit || tierLimits.chat_messages
    },
    api_calls: {
      current: user.api_calls_count || 0,
      limit: user.api_calls_limit || tierLimits.api_calls
    },
    storage_bytes: {
      current: user.storage_used_bytes || 0,
      limit: user.storage_limit_bytes || tierLimits.storage_bytes
    }
  }

  const { current, limit } = limitMapping[limitType]
  const wouldExceed = (current + increment) > limit

  return {
    withinLimits: !wouldExceed,
    usageData: {
      tier,
      current,
      limit,
      would_exceed: wouldExceed,
      remaining: Math.max(0, limit - current)
    }
  }
}

export async function enforceUsageLimit(
  userId: string,
  limitType: LimitType,
  increment: number = 1
): Promise<UsageData> {
  const { withinLimits, usageData } = await checkUsageLimit(userId, limitType, increment)

  if (!withinLimits) {
    throw new UsageLimitError(
      limitType,
      usageData.current,
      usageData.limit,
      usageData.tier,
      usageData.tier === 'free'
    )
  }

  return usageData
}

export async function incrementUsageCounter(
  userId: string,
  eventType: 'document_upload' | 'chat_message' | 'api_call',
  resourceCount: number = 1,
  eventData: Record<string, unknown> = {}
): Promise<boolean> {
  try {
    // Try to use database function first (if available) in app schema
    const { data, error } = await supabaseApp.rpc('increment_usage', {
      p_user_id: userId,
      p_event_type: eventType,
      p_resource_count: resourceCount,
      p_event_data: eventData
    })

    if (!error) {
      return data || true
    }

    // Fallback to manual increment if RPC function doesn't exist
    console.warn('RPC function not available, using manual increment')
    return await manualIncrementUsage(userId, eventType, resourceCount)

  } catch (error) {
    console.error(`Error incrementing usage for ${userId}:`, error)
    // Try manual increment as fallback
    return await manualIncrementUsage(userId, eventType, resourceCount)
  }
}

async function manualIncrementUsage(
  userId: string,
  eventType: 'document_upload' | 'chat_message' | 'api_call',
  resourceCount: number
): Promise<boolean> {
  try {
    const user = await getUserById(userId)
    if (!user) return false

    const updates: Partial<User> = {}

    switch (eventType) {
      case 'document_upload':
        updates.documents_uploaded = (user.documents_uploaded || 0) + resourceCount
        break
      case 'chat_message':
        updates.chat_messages_count = (user.chat_messages_count || 0) + resourceCount
        break
      case 'api_call':
        updates.api_calls_count = (user.api_calls_count || 0) + resourceCount
        break
    }

    await updateUser(userId, updates)
    return true

  } catch (error) {
    console.error(`Error in manual increment for ${userId}:`, error)
    return false
  }
}

export async function checkAndIncrementUsage(
  userId: string,
  limitType: LimitType,
  eventType: 'document_upload' | 'chat_message' | 'api_call',
  increment: number = 1,
  eventData: Record<string, unknown> = {}
): Promise<UsageData> {
  // First check if within limits
  const usageData = await enforceUsageLimit(userId, limitType, increment)

  // If within limits, increment the counter
  const success = await incrementUsageCounter(userId, eventType, increment, eventData)

  if (!success) {
    console.error(`Failed to increment usage counter for ${userId}`)
  }

  return usageData
}

export function getUsageWarningThreshold(tier: SubscriptionTier, limitType: LimitType): number {
  const tierLimits = TIER_LIMITS[tier]
  if (!tierLimits || !(limitType in tierLimits)) {
    return 0
  }

  const limit = tierLimits[limitType]
  // Warn at 80% of limit
  return Math.floor(limit * 0.8)
}

export function formatStorageSize(bytesSize: number): string {
  const units = ['B', 'KB', 'MB', 'GB', 'TB']
  let size = bytesSize
  let unitIndex = 0

  while (size >= 1024 && unitIndex < units.length - 1) {
    size /= 1024
    unitIndex++
  }

  return `${size.toFixed(1)} ${units[unitIndex]}`
}

export function getUpgradeMessage(tier: SubscriptionTier): string {
  switch (tier) {
    case 'free':
      return 'Upgrade to Pro for higher limits and better performance.'
    case 'pro':
      return 'Consider Pro BYOK for even higher limits with your own API keys.'
    case 'pro_byok':
      return "You're on our highest tier. Contact support for enterprise options."
    default:
      return 'Upgrade your plan for higher limits.'
  }
}

export async function resetMonthlyUsage(): Promise<boolean> {
  try {
    const { error } = await supabaseApp.rpc('reset_monthly_usage')
    if (error) {
      console.error('Error resetting monthly usage:', error)
      return false
    }
    console.info('Monthly usage reset completed')
    return true
  } catch (error) {
    console.error('Error resetting monthly usage:', error)
    return false
  }
}

export async function getUsageAnalytics(days: number = 30): Promise<unknown[]> {
  try {
    const { data, error } = await supabaseApp
      .from('usage_analytics')
      .select('*')
      .limit(days * 10)

    if (error) {
      console.error('Error fetching usage analytics:', error)
      return []
    }

    return data || []
  } catch (error) {
    console.error('Error fetching usage analytics:', error)
    return []
  }
}

// Usage logging for analytics in app schema
export async function logUsage(
  userId: string,
  action: string,
  details: Record<string, unknown> = {},
  costCents: number = 0
): Promise<boolean> {
  try {
    const { error } = await supabaseApp
      .from('usage_logs')
      .insert({
        user_id: userId,
        action,
        details,
        cost_cents: costCents
      })

    if (error) {
      console.error('Error logging usage:', error)
      return false
    }

    return true
  } catch (error) {
    console.error('Error logging usage:', error)
    return false
  }
}
